src/main/java/net/sf/jabref/remote/DBSynchronizer.java

```diff
@@ -0,0 +1,227 @@
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+package net.sf.jabref.remote;
+
+import java.sql.Connection;
+import java.util.List;
+import java.util.Set;
+
+import net.sf.jabref.MetaData;
+import net.sf.jabref.event.EntryAddedEvent;
+import net.sf.jabref.event.EntryEvent;
+import net.sf.jabref.event.EntryRemovedEvent;
+import net.sf.jabref.event.FieldChangedEvent;
+import net.sf.jabref.event.MetaDataChangedEvent;
+import net.sf.jabref.event.location.EntryEventLocation;
+import net.sf.jabref.exporter.BibDatabaseWriter;
+import net.sf.jabref.logic.l10n.Localization;
+import net.sf.jabref.model.database.BibDatabase;
+import net.sf.jabref.model.entry.BibEntry;
+
+import com.google.common.eventbus.Subscribe;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * Synchronizes the remote or local databases with their opposite side.
+ * Local changes are pushed by {@link EntryEvent} using Google's Guava EventBus.
+ */
+public class DBSynchronizer {
+
+    private static final Log LOGGER = LogFactory.getLog(DBConnector.class);
+
+    private DBProcessor dbProcessor;
+    private DBType dbType;
+    private String dbName;
+    private MetaData metaData;
+    private final BibDatabase bibDatabase;
+
+    public DBSynchronizer(BibDatabase bibDatabase, MetaData metaData) {
+        this.bibDatabase = bibDatabase;
+        this.metaData = metaData;
+    }
+
+    /**
+     * Listening method. Inserts a new {@link BibEntry} remotely.
+     * @param event {@link EntryAddedEvent} object
+     */
+    @Subscribe
+    public void listen(EntryAddedEvent event) {
+        // While synchronizing the local database (see synchronizeLocalDatabase() below), some EntryEvents may be posted.
+        // In this case DBSynchronizer should not try to insert the bibEntry entry again (but it would not harm).
+        if (isInEventLocation(event)) {
+            dbProcessor.insertEntry(event.getBibEntry());
+            synchronizeLocalMetaData();
+            synchronizeLocalDatabase(); // Pull remote changes for the case that there where some
+        }
+    }
+
+    /**
+     * Listening method. Updates an existing remote {@link BibEntry}.
+     * @param event {@link FieldChangedEvent} object
+     */
+    @Subscribe
+    public void listen(FieldChangedEvent event) {
+        // While synchronizing the local database (see synchronizeLocalDatabase() below), some EntryEvents may be posted.
+        // In this case DBSynchronizer should not try to update the bibEntry entry again (but it would not harm).
+        if (isInEventLocation(event)) {
+            synchronizeLocalMetaData();
+            BibDatabaseWriter.applySaveActions(event.getBibEntry(), metaData);
```

Comment:
Yes they are. This makes sure that data which is going to be pushed is formatted according to the meta data, which is always up-to-date.